jsapp/js/components/account/mfaModals.tsx

```diff
@@ -0,0 +1,487 @@
+
+    switch (this.props.modalType) {
+      case 'regenerate':
+        this.mfaRegenerate()
+        break
+      case 'reconfigure':
+        this.mfaDeactivate()
+        break
+      case 'deactivate':
+        this.mfaDeactivate()
+        break
+    }
+  }
+
+  onInputChange(inputString: string) {
+    this.setState({inputString: inputString})
+  }
+
+  changeStep(
+    evt: React.ChangeEvent<HTMLInputElement>,
+    nextStep: modalSteps
+  ) {
+    evt.preventDefault()
+
+    this.setState({currentStep: nextStep})
+  }
+
+  isTokenValid(): boolean {
+    return this.state.inputString !== null && this.state.inputString.length === 6
+  }
+
+  downloadCodes() {
+    if (this.state.backupCodes) {
+      const USERNAME = stores.session.currentAccount.username
+      // gets date in yyyymmdd
+      const DATE = new Date().toJSON().slice(0, 10).replace(/-/g, '')
+
+      const formatedCodes = this.state.backupCodes.map((t) => {
+        return t + '\n'
+      })
+      const codesLink = document.createElement('a')
+      const codesFile = new Blob(formatedCodes)
+
+      codesLink.href = URL.createObjectURL(codesFile)
+      codesLink.download = 'backups_' + USERNAME + '_' + DATE + '.txt'
+
+      document.body.appendChild(codesLink)
+      codesLink.click()
+
+      this.setState({downloadClicked: true})
+    }
+  }
+
+  // HACK FIX: since the header is seperate from the modal we do this
+  // roundabout way of disabling the close icon
+  disableCloseIcon() {
+    const closeIcon = document.getElementsByClassName('modal__x')[0] as HTMLElement
+    closeIcon.hidden = true
+  }
+
+  renderQRCodeStep() {
+    return (
+      <bem.MFAModals__qrstep>
+        <bem.MFAModals__description>
+          {t('Two-factor Authenication (2FA) is an added layer of security used when logging into the platform. We reccomend enabling Two-factor Authenication for an additional layer of protection.')}
+        </bem.MFAModals__description>
+
+        <bem.MFAModals__body>
+          <bem.MFAModals__qr>
+            <QRCode
+              value={this.state.qrCode || ''}
+              size={240}
+            />
+          </bem.MFAModals__qr>
+
+          <bem.MFAModals__token>
+            <strong>
+              {t('Scan QR code and enter the six-digit token from the application')}
+            </strong>
+
+            {t('After scanning the QR code image, the app will display a six-digit code that you can display below.')}
+
+            <TextBox
+              errors={this.state.errorText}
+              value={this.state.inputString}
+              onChange={this.onInputChange.bind(this)}
+              customModifiers={'on-white'}
+            />
+            <bem.MFAModals__manual>
+              {t('No QR code?')}
+
+              &nbsp;
+
+              <bem.MFAModals__manual__link
+                onClick={(evt: React.ChangeEvent<HTMLInputElement>) => {
+                  this.changeStep(evt, 'manual')
+                }}
```

Comment:
II thought it would too but it looks like typescript doesn't like it, I get this nasty blob when I try it:
![Screenshot from 2022-02-17 16-24-36](https://user-images.githubusercontent.com/16762675/154573302-6c5f3bfa-29e7-49b6-ab15-3cad845eefbb.png)